core/src/main/java/org/infinispan/container/offheap/OffHeapDataContainer.java

```diff
@@ -151,10 +152,11 @@ protected void checkDeallocation() {
 
    protected InternalCacheEntry<WrappedBytes, WrappedBytes> performGet(long address, Object k, boolean peek) {
       WrappedBytes wrappedKey = toWrapper(k);
+      long now = timeService.wallClockTime();
       while (address != 0) {
          long nextAddress = offHeapEntryFactory.getNext(address);
          InternalCacheEntry<WrappedBytes, WrappedBytes> ice = offHeapEntryFactory.fromMemory(address);
-         if (wrappedKey.equalsWrappedBytes(ice.getKey())) {
+         if (wrappedKey.equalsWrappedBytes(ice.getKey()) && !ice.isExpired(now)) {
```

Comment:
This can be short circuited as well so we don't have to iterate through all addresses.